k9mail-library/src/test/java/com/fsck/k9/mail/filter/LineWrapOutputStreamTest.java

```diff
@@ -0,0 +1,308 @@
+        for (int i = 0; i < MAX_BYTES_PER_LINE; ++i) {
+            if (nextWordPos <= i) {
+                writer.write(' ');
+                nextWordPos += 2 + rand.nextInt(MAX_WORD_LEN);
+            } else {
+                int nextByte = rand.nextInt(64000);
+                while (nextByte == '\n' || nextByte == '\r') nextByte = rand.nextInt(64000);
+                writer.write(nextByte);
+            }
+        }
+    }
+
+    /**
+     * Random bytes because it is hard to count the bytes in preset strings
+     * which are the datums on which wrapping occurs.
+     * CRLF characters are filtered from the input since we will be
+     * assessing optimal, non-overflowing, wrapping and empty, excessive, breaks
+     * would interfere.
+     * @throws Exception
+     */
+    @Test
+    public void testTightWrappingWithRandBytes() throws Exception {
+        for (int i = 0; i < 10; ++i) {
+            testTightWrapRandom(20, baos, rand);
+            testTightWrapRandom(40, baos, rand);
+            testTightWrapRandom(60, baos, rand);
+            testTightWrapRandom(80, baos, rand);
+            testTightWrapRandom(100, baos, rand);
+            testTightWrapRandom(120, baos, rand);
+            testTightWrapRandom(240, baos, rand);
+            testTightWrapRandom(480, baos, rand);
+            testTightWrapRandom(1024, baos, rand);
+            testTightWrapRandom(2048, baos, rand);
+        }
+    }
+
+    private static void checkWrapsTight(int MAX_BYTES_PER_LINE, ByteArrayOutputStream baos) {
+        byte bArray[] = baos.toByteArray();
+        int bytesRead = 0;
+        int lastLinesLen = -1;
+        int lineStartPos = 0;
+        int lastLineStartPos = -1;
+        do {
+            int firstSpace = -1;
+            for (int i = 0; i < MAX_BYTES_PER_LINE && bytesRead < bArray.length; ++i) {
+                byte curByte = bArray[bytesRead++];
+                if (curByte == '\r') {
+                    lastLinesLen = i;
+                    lastLineStartPos = lineStartPos;
+                    lineStartPos = ++bytesRead;
+                    break;
+                }
+                if (curByte == ' ') {
+                    if (firstSpace < 0) {
+                        firstSpace = i;
+                        if (lastLinesLen >= 0) {
+                            //Configuration for a 20 byte line will only attempt 17 + CRLF.
+                            final int WASTED_LEN = 1;
+                            if (MAX_BYTES_PER_LINE - 3 - WASTED_LEN - lastLinesLen >= firstSpace) {
+                                String lastString = new String(bArray, lastLineStartPos
+                                        , lineStartPos - lastLineStartPos - 2);
+                                String curString = new String(bArray, lineStartPos
+                                        , i);
+                                try {
+                                    System.out.println(new String(bArray, "UTF-8"));
```

Comment:
The fail message appears directly after, think I was meaning to integrate the two.